### PR TITLE
[MM-47896] Channel names can collide with GM names

### PR DESCRIPTION
**Pull request by mylonsuren**
#### Summary

Blocks channel creation if name matches the GM naming pattern `/[0-9a-f]{40}/`.

#### Ticket Link

[MM-47896](https://mattermost.atlassian.net/browse/MM-47896)

#### Release Note

```release-note
Fixed issue where new channel names could collide with existing GM names.
```

### CODE REVIEW

**Comment by mylonsuren** [comment]
@esethna Just wanted to loop you in as I changed the text that is shown to the user if they run into this. Formerly it was just "User ids are not permitted in channel name", now it is "User and Channel ids are not permitted in channel name". Let me know if this is all good or if we should have something different for this error.

**Comment by esethna** [comment]
@mylonsuren I'll pass this text review to @cwarnermm. Thanks for looping us in.

**Comment by cwarnermm** [comment]
Thanks for including me, @esethna. Because this is a security-driven update, I strongly recommend that the error say only: ``Invalid channel name.``

Offering more context could provide bad actors with too much information they could try to exploit. We want to avoid any additional product text that could confirm whether something's valid/in-use, etc. A bad actor that hits this error and receives confirmation that it's a duplicate channel name would know the ID of an existing group message channel.

File: model/channel.go [hunk 1]
```diff
@@ -281,9 +282,11 @@ func (o *Channel) IsValid() *AppError {
 		return NewAppError("Channel.IsValid", "model.channel.is_valid.creator_id.app_error", nil, "", http.StatusBadRequest)
 	}
 
-	userIds := strings.Split(o.Name, "__")
-	if o.Type != ChannelTypeDirect && len(userIds) == 2 && IsValidId(userIds[0]) && IsValidId(userIds[1]) {
-		return NewAppError("Channel.IsValid", "model.channel.is_valid.name.app_error", nil, "", http.StatusBadRequest)
+	if o.Type != ChannelTypeDirect && o.Type != ChannelTypeGroup {
+		userIds := strings.Split(o.Name, "__")
+		if ok, _ := regexp.MatchString("^[a-fA-F0-9]{40}$", o.Name); ok || o.Type != ChannelTypeDirect && len(userIds) == 2 && IsValidId(userIds[0]) && IsValidId(userIds[1]) {
```
[the remark this answers is not in the document]
Comment (mylonsuren) [reply model/channel.go]: Done 😄

**Comment by mylonsuren** [comment]
> Because this is a security-driven update, I strongly recommend that the error say only: `Invalid channel name.`
> 
> Offering more context could provide bad actors with too much information they could try to exploit. We want to avoid any additional product text that could confirm whether something's valid/in-use, etc. A bad actor that hits this error and receives confirmation that it's a duplicate channel name would know the ID of an existing group message channel.

Thanks @cwarnermm! I've updated the text to just `Invalid channel name.`

File: model/channel.go [hunk 2]
```diff
@@ -281,9 +284,11 @@ func (o *Channel) IsValid() *AppError {
 		return NewAppError("Channel.IsValid", "model.channel.is_valid.creator_id.app_error", nil, "", http.StatusBadRequest)
 	}
 
-	userIds := strings.Split(o.Name, "__")
-	if o.Type != ChannelTypeDirect && len(userIds) == 2 && IsValidId(userIds[0]) && IsValidId(userIds[1]) {
-		return NewAppError("Channel.IsValid", "model.channel.is_valid.name.app_error", nil, "", http.StatusBadRequest)
+	if o.Type != ChannelTypeDirect && o.Type != ChannelTypeGroup {
+		userIds := strings.Split(o.Name, "__")
+		if ok := gmNameRegex.MatchString(o.Name); ok || o.Type != ChannelTypeDirect && len(userIds) == 2 && IsValidId(userIds[0]) && IsValidId(userIds[1]) {
```
Comment (ashishbhate) [inline model/channel.go:289]: `o.Type != ChannelTypeDirect && len(userIds) == 2 && IsValidId(userIds[0]) && IsValidId(userIds[1])`

I know this part of the condition wasn't added in this PR, but it looks like its not covered in the tests. Could we add a test for this condition as well, to lock in the behaviour? It should only be two more lines:

```
o.Name = "71b03afcbb2d503d49f87f057549c43db4e19f92__pazfdvmko08qhzd0icg76zo9gr3w42ah05"
require.NotNil(t, o.IsValid())
```

**Review by ashishbhate — APPROVED** [review]

File: model/channel.go (hunk 2 above)
Comment (mylonsuren) [reply model/channel.go]: Good call 👍

**Comment by jupenur** [comment]
> A bad actor that hits this error and receives confirmation that it's a duplicate channel name would know the ID of an existing group message channel.

Good thinking @cwarnermm but this isn't actually a concern in this case. The code doesn't check whether the name is a real GM channel name; it only checks whether it "looks" like one, i.e. is the same length and is made up of the same character set. No sensitive information would leak even with a more descriptive message.

**Review by jupenur — APPROVED** [review]

**Comment by esethna** [comment]
> What do you think, @esethna? Do we want to go so far as to say that channel names can't be of a specific format, or is it sufficient to ask the user to choose a different name without telling the user why?

@cwarnermm your call on the level of detail, but I agree with offering the path to success: ie Please enter a different channel name

**Comment by agnivade** [comment]
No need for my review @mylonsuren. It was just a drive by comment :)

**Comment by cwarnermm** [comment]
Thanks, @esethna. @mylonsuren - Please update the text string to: ``Channel names can't be in a hexadecimal format. Please enter a different channel name.``

Not every user will understand what ``hexadecimal format`` means; however, if we start seeing questions about it, that will tell us something about how often users are seeing this error message (and entering channel names like this), which will be insightful.

**Comment by jgilliam17** [comment]
Thanks @mylonsuren 
Few tests are failing due to text change see full E2E report [here ](https://automation-dashboard.vercel.app/cycles/7429ad17-1eb8-43a8-b22d-5463f7ca2f4e)

**Comment by mylonsuren** [comment]
> Thanks @mylonsuren Few tests are failing due to text change see full E2E report [here ](https://automation-dashboard.vercel.app/cycles/7429ad17-1eb8-43a8-b22d-5463f7ca2f4e)

Sorry for the delay @jgilliam17, link doesn't work for me anymore, but I'm guessing it was `MM-T884_2` that was failing?

**Comment by jgilliam17** [comment]
Thanks @mylonsuren for re-running e2es - here is the latest [report](https://mattermost-cypress-report.s3.amazonaws.com/269708-dc97ac9-onprem-ent-server-pr-21639-mm-ee-test:dc97ac9/mochawesome.html)
MM-T884_2
MM-T883 

Additional tests that are failing (and also differ from master), but don't appear to be related to PR
MM-T4943_1
MM-T4943_2
MM-44435

**Comment by mylonsuren** [comment]
@jgilliam17 I've created a new web app PR https://github.com/mattermost/mattermost-webapp/pull/11716 to address the failing tests.

**Comment by jgilliam17** [comment]
@mylonsuren Can you please fix failing checks so i can trigger e2es again? Thanks 🙂

**Comment by jgilliam17** [comment]
There is one check pending since yesterday. Can you please take a look why this is happening? @mylonsuren

**Comment by mylonsuren** [comment]
@jgilliam17 Failing checks have been fixed, should be good for you to take a look at 😄

**Comment by jgilliam17** [comment]
Thanks @mylonsuren new [e2e report ](https://mattermost-cypress-report.s3.amazonaws.com/278916-58657b2-onprem-ent-server-pr-21639-mm-ee-test:58657b2/mochawesome.html) looks good. Thanks for filing a follow up to add new test.

**Review by jgilliam17 — APPROVED** [review]

**Comment by mylonsuren** [comment]
Awaiting the associated webapp PR with the E2E tests to be approved before merging this.

**Pull request merged**